File: app/src/main/java/com/paxus/pay/poslinkui/demo/utils/InterfaceHistory.java
package com.paxus.pay.poslinkui.demo.utils;

import android.annotation.SuppressLint;

import androidx.annotation.NonNull;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.ListIterator;
import java.util.Stack;

public class InterfaceHistory {

    private Stack<InterfaceRecord> history;

    public InterfaceHistory(){
        history = new Stack<>();
        Logger.d("Interface History Initialized.");
    }

    public void add(@NonNull String interfaceID, @NonNull String interfaceAction) {
        InterfaceRecord newInterfaceRecord = new InterfaceRecord(interfaceID, interfaceAction, System.currentTimeMillis());
        Logger.d("New Interface: " + newInterfaceRecord);
        history.push(newInterfaceRecord);
    }

    public boolean validate(String intendedInterfaceID, String intendedInterfaceAction) {
        if(history.empty()) return true;

        boolean isInterfaceIDValid = intendedInterfaceID != null && history.peek().interfaceID.equals(intendedInterfaceID);
        boolean isInterfaceActionValid = intendedInterfaceAction != null && history.peek().interfaceAction.equals(intendedInterfaceAction);
        if (isInterfaceIDValid && isInterfaceActionValid) history.peek().accept();
        print();
        return isInterfaceIDValid && isInterfaceActionValid;
    }

    private void print() {
        StringBuilder historyString = new StringBuilder(getClass().getSimpleName() + "\n");
        ListIterator<InterfaceRecord> iterator = history.listIterator(history.size());
        while(iterator.hasPrevious()) historyString.append(iterator.previous().toString());
        Logger.d(historyString.toString());
    }

    private class InterfaceRecord {
        public String interfaceID, interfaceAction;
        public long initTimeMillis;
        private boolean isAccepted = false;

        public InterfaceRecord(@NonNull String interfaceID, @NonNull String interfaceAction, long initTimeMillis) {
            this.interfaceID = interfaceID;
            this.interfaceAction = interfaceAction;
            this.initTimeMillis = initTimeMillis;
        }

        public void accept() {
            this.isAccepted = true;
        }

        @NonNull @Override @SuppressLint("SimpleDateFormat") public String toString() {
            return new SimpleDateFormat("HH:mm:ss:SSS").format(new Date(this.initTimeMillis)) +
                    ":  " + this.interfaceAction + (this.isAccepted ? " (Accepted)" : "") + "\t\t" + this.interfaceID + "\n";
        }
    }
}

File: app/src/main/java/com/paxus/pay/poslinkui/demo/utils/InterfaceHistory.kt
package com.paxus.pay.poslinkui.demo.utils

import android.annotation.SuppressLint
import javax.inject.Inject
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Stack

class InterfaceHistory @Inject constructor() {
    private val history: Stack<InterfaceRecord?>

    init {
        history = Stack<InterfaceRecord?>()
        Logger.d("Interface History Initialized.")
    }

    fun add(interfaceID: String, interfaceAction: String) {
        val newInterfaceRecord =
            InterfaceRecord(interfaceID, interfaceAction, System.currentTimeMillis())
        Logger.d("New Interface: " + newInterfaceRecord)
        history.push(newInterfaceRecord)
    }

    fun validate(intendedInterfaceID: String?, intendedInterfaceAction: String?): Boolean {
        if (history.empty()) return true

        val top = history.peek() ?: return true
        val isInterfaceIDValid =
            intendedInterfaceID != null && top.interfaceID == intendedInterfaceID
        val isInterfaceActionValid =
            intendedInterfaceAction != null && top.interfaceAction == intendedInterfaceAction
        if (isInterfaceIDValid && isInterfaceActionValid) top.accept()
        print()
        return isInterfaceIDValid && isInterfaceActionValid
    }

    private fun print() {
        val historyString = StringBuilder(javaClass.getSimpleName() + "\n")
        val iterator = history.listIterator(history.size)
        while (iterator.hasPrevious()) historyString.append(iterator.previous().toString())
        Logger.d(historyString.toString())
    }

    private inner class InterfaceRecord(
        var interfaceID: String,
        var interfaceAction: String,
        var initTimeMillis: Long
    ) {
        private var isAccepted = false

        fun accept() {
            this.isAccepted = true
        }

        @SuppressLint("SimpleDateFormat")
        override fun toString(): String {
            return SimpleDateFormat("HH:mm:ss:SSS").format(Date(this.initTimeMillis)) +
                    ":  " + this.interfaceAction + (if (this.isAccepted) " (Accepted)" else "") + "\t\t" + this.interfaceID + "\n"
        }
    }
}
